File: roam_io/functions/party_season.js
/*
 * Description:
 *   Fortnightly Party Mode season reset: summarizes tile ownership into a
 *   persisted season record, then wipes tile ownership/dwell counters so the
 *   next season starts clean.
 */

const { deriveOwnership } = require('./party_dwell');

async function resetPartySeason({ db, partyId, now }) {
  const partyRef = db.collection('parties').doc(partyId);
  const partyDoc = await partyRef.get();
  const party = partyDoc.data();

  const tilesCollection = partyRef.collection('tiles');
  const tilesSnapshot = await tilesCollection.get();

  let teamATileCount = 0;
  let teamBTileCount = 0;
  for (const doc of tilesSnapshot.docs) {
    const owner = deriveOwnership(doc.data());
    if (owner === 'A') teamATileCount += 1;
    if (owner === 'B') teamBTileCount += 1;
  }

  const winner =
    teamATileCount === teamBTileCount
      ? null
      : teamATileCount > teamBTileCount
        ? 'A'
        : 'B';

  const nowIso = now.toISOString();
  const summary = {
    teamAMembers: party.teamAMembers,
    teamBMembers: party.teamBMembers,
    teamATileCount,
    teamBTileCount,
    winner,
    startAt: party.currentSeasonStartAt,
    endAt: nowIso,
  };

  await partyRef.collection('seasons').doc().set(summary);

  for (const doc of tilesSnapshot.docs) {
    await doc.ref.delete();
  }
  await partyRef.set(
    { ...party, currentSeasonStartAt: nowIso },
  );

  return summary;
}

module.exports = { resetPartySeason };

File: roam_io/lib/features/party/domain/party_season.dart
/// A persisted Party Mode season summary (no per-tile/per-user breakdown).
class PartySeason {
  const PartySeason({
    required this.id,
    required this.teamAMembers,
    required this.teamBMembers,
    required this.teamATileCount,
    required this.teamBTileCount,
    required this.winner,
    required this.startAt,
    required this.endAt,
  });

  final String id;
  final List<String> teamAMembers;
  final List<String> teamBMembers;
  final int teamATileCount;
  final int teamBTileCount;
  final String? winner;
  final DateTime? startAt;
  final DateTime endAt;

  factory PartySeason.fromMap(String id, Map<String, dynamic> map) {
    return PartySeason(
      id: id,
      teamAMembers: List<String>.from(map['teamAMembers'] as List? ?? const []),
      teamBMembers: List<String>.from(map['teamBMembers'] as List? ?? const []),
      teamATileCount: map['teamATileCount'] as int? ?? 0,
      teamBTileCount: map['teamBTileCount'] as int? ?? 0,
      winner: map['winner'] as String?,
      startAt: map['startAt'] != null
          ? DateTime.parse(map['startAt'] as String)
          : null,
      endAt: DateTime.parse(map['endAt'] as String),
    );
  }
}
